### DISCUSSION TITLE
TOWER FORE-AFT MODE SHAPES, coefficients of polynomial terms

### DISCUSSION THREAD

**Discussion by EWEMkite** [OpenFAST/openfast#2551] [Q&A]
Dear @jjonkman,

I write a python code for the modal analysis of the tower, in the code I establish the mass and stiffness matrix then use  `eigenvalues, eigenvectors = eigh(stiffness_matrix, mass_matrix_scaled)`. But because the coordinates are defined as (z/L), the eigenvectors [6X6] give the coefficients of (z/L) ~ (z/L)^6 for mode 1~6 I think. As far as I know, the mode shape and eigenvector are connected by for examle: mode shape [Z] = sum(eigenvector[:, 0] · [(z/L), (z/L)^2, (z/L)^3, (z/L)^4, (z/L)^5, (z/L)^6])

I am not very sure what should I do to get the required coefficients for x^1 ~ x^6 in ElastoDyn file, should I just multiply eigenvectors with L ~ L^6, or should I just use z rather than z/L when building the matrix? I use z/L because it's an equation I found in this paper:
https://www.researchgate.net/publication/372289833_A_Rayleigh-Ritz_solution_for_high_order_natural_frequencies_and_eigenmodes_of_monopile_supported_offshore_wind_turbines_considering_tapered_towers_and_soil_pile_interactions

**Comment by jjonkman** [comment]
Dear @EWEMkite,

I'm not really sure I understand your question.  If you are looking to fit the mode shape polynomial for ElastoDyn from a given mode shape, the old _ModeShapePolyFitting.xls_ spreadsheet would compute these using a number of different methods.  A MATLAB-based version of this script was provided on our forum: https://forums.nrel.gov/t/fast-tower-polynomial-mode-shapes-from-bmodes-matlab-script/1637.  If you are looking for a Python-based tool, while I haven't used this myself, I found the `_get_modal_coefficients()` function of the _bmodes_out_file.py_ script in the OpenFAST Toolbox: https://github.com/OpenFAST/openfast_toolbox/blob/main/openfast_toolbox/io/bmodes_out_file.py.

Best regards,

**Reply by EWEMkite** [reply]
Dear @jjonkman ,

I tried with BModes, but the outputs show a very small frequency (the first mode has a magnitude of 0.0001). I suppose there is something wrong in the input files, could you help me have a check? Thanks.
`======================   BModes v1.03 Main Input File  ==================
Modes of a 82.4m tower with tip mass

--------- General parameters ---------------------------------------------------------------------
False     Echo        Echo input file contents to *.echo file if true.
2         beam_type   1: blade, 2: tower (-)
0.        romg:       rotor speed (rpm), automatically set to zero for tower modal analysis
1.0       romg_mult:  rotor speed muliplicative factor (-)
82.39     radius:     rotor tip radius measured along coned blade axis OR tower height (m)
0.        hub_rad:    hub radius measured along coned blade axis OR tower rigid-base height (m)
0.        precone:    built-in precone angle (deg), automatically set to zero for a tower
0.        bl_thp:     blade pitch setting (deg), automatically set to zero for a tower
1         hub_conn:   hub-to-blade connection [1: cantilevered; other options not yet available]
20        modepr:     number of modes to be printed (-)
f         TabDelim    (true: tab-delimited output tables; false: space-delimited tables)
f         mid_node_tw  (true: output twist at mid-node of elements; false: no mid-node outputs)

--------- Blade-tip or tower-top mass properties --------------------------------------------
1026410.750   tip_mass    blade-tip or tower-top mass (see users' manual) (kg)
-3.3      cm_loc      tip-mass c.m. offset from the tower axis measured along the tower-tip x reference axis (m)
1.1       cm_axial    tip-mass c.m. offset tower tip measures axially along the z axis (m)
6.22E+09  ixx_tip     blade lag or tower s-s mass moment of inertia about the tip-section x reference axis (kg-m^2)
6.22E+09  iyy_tip     blade flap or tower f-a mass moment of inertia about the tip-section y reference axis (kg-m^2)
2.66E+07  izz_tip     torsion mass moment of inertia about the tip-section z reference axis (kg-m^2)
0.        ixy_tip     cross product of inertia about x and y reference axes(kg-m^2)
-18400.   izx_tip     cross product of inertia about z and x reference axes(kg-m^2)
0.        iyz_tip     cross product of inertia about y and z reference axes(kg-m^2)

--------- Distributed-property identifiers --------------------------------------------------------
1         id_mat:     material_type [1: isotropic; non-isotropic composites option not yet available]
'IEA15_CSpar_Bmodes.dat' sec_props_file   name of beam section properties file (-)

Property scaling factors..............................
1.0       sec_mass_mult:   mass density multiplier (-)
1.0       flp_iner_mult:   blade flap or tower f-a inertia multiplier (-)
1.0       lag_iner_mult:   blade lag or tower s-s inertia multiplier (-)
1.0       flp_stff_mult:   blade flap or tower f-a bending stiffness multiplier (-)
1.0       edge_stff_mult:  blade lag or tower s-s bending stiffness multiplier (-)
1.0       tor_stff_mult:   torsion stiffness multiplier (-)
1.0       axial_stff_mult: axial stiffness multiplier (-)
1.0       cg_offst_mult:   cg offset multiplier (-)
1.0       sc_offst_mult:   shear center multiplier (-)
1.0       tc_offst_mult:   tension center multiplier (-)

--------- Finite element discretization --------------------------------------------------
20        nselt:     no of blade or tower elements (-)
Distance of element boundary nodes from blade or flexible-tower root (normalized wrt blade or tower length), el_loc()
0.00 	0.05 	0.10 	0.15 	0.20 	0.25 	0.30 	0.35 	0.40 	0.45 	0.50 	0.55 	0.60 	0.65 	0.70 	0.75 	0.80 	0.85 	0.90 	0.95 	1.00 

--------- Properties of additional tower support subsystem (read only if beam_type is 2) ------------
0         tow_support: : additional tower support [0: no additional support; 1: Tension guy wires for land-based tower; 2: offshore turbine support: floating platform or monopile] (-)
Tension-wires data
0         n_attachments: no of wire-attachment locations on tower, maxm allowable is 2; 0: no tension-wire support (-)
3 3       n_wires:       no of wires attached at each location (must be 3 or higher) (-)
6 9       node_attach:   node numbers of attacments location (node number must be more than 1 and less than nselt+2) (-)
1.e8 1.e8 wire_stfness:  wire sifnness in each set (see users' manual) (N/m)
45. 45.   th_wire:       angle of tension wires (wrt a horizontal plane) at each attachment point (deg)
`
`Tower section properties (WindPact 1.5 MW turbine)
21        n_secs:     number of blade or tower sections at which properties are specified (-)

sec_loc	str_tw	tw_iner	mass_den	flp_iner	edge_iner	flp_stff	edge_stff	tor_stff	axial_stff	cg_offst	sc_offst	tc_offst
(-)	(deg)	(deg)	(kg/m)	(kg-m)	(kg-m)	(Nm^2)	(Nm^2)	(Nm^2)	(N)	(m)	(m)	(m)
0		0	0	1.03E+04	2.14E+05	2.14E+05	4.59E+12	4.59E+12	1.03E+03	2.28E+02	0	0	0
0.05	0	0	1.01E+04	1.98E+05	1.98E+05	4.25E+12	4.25E+12	9.52E+02	2.23E+02	0	0	0
0.1		0	0	9.80E+03	1.83E+05	1.83E+05	3.93E+12	3.93E+12	8.80E+02	2.17E+02	0	0	0
0.15	0	0	9.54E+03	1.69E+05	1.69E+05	3.63E+12	3.63E+12	8.12E+02	2.11E+02	0	0	0
0.2		0	0	9.28E+03	1.55E+05	1.55E+05	3.34E+12	3.34E+12	7.47E+02	2.05E+02	0	0	0
0.25	0	0	9.02E+03	1.43E+05	1.43E+05	3.06E+12	3.06E+12	6.86E+02	2.00E+02	0	0	0
0.3		0	0	8.76E+03	1.31E+05	1.31E+05	2.81E+12	2.81E+12	6.28E+02	1.94E+02	0	0	0
0.35	0	0	8.50E+03	1.19E+05	1.19E+05	2.56E+12	2.56E+12	5.74E+02	1.88E+02	0	0	0
0.4		0	0	8.24E+03	1.09E+05	1.09E+05	2.33E+12	2.33E+12	5.22E+02	1.82E+02	0	0	0
0.45	0	0	7.98E+03	9.87E+04	9.87E+04	2.12E+12	2.12E+12	4.74E+02	1.76E+02	0	0	0
0.5		0	0	7.72E+03	8.93E+04	8.93E+04	1.92E+12	1.92E+12	4.29E+02	1.71E+02	0	0	0
0.55	0	0	7.46E+03	8.06E+04	8.06E+04	1.73E+12	1.73E+12	3.87E+02	1.65E+02	0	0	0
0.6		0	0	7.20E+03	7.24E+04	7.24E+04	1.56E+12	1.56E+12	3.48E+02	1.59E+02	0	0	0
0.65	0	0	6.94E+03	6.48E+04	6.48E+04	1.39E+12	1.39E+12	3.12E+02	1.53E+02	0	0	0
0.7		0	0	6.68E+03	5.78E+04	5.78E+04	1.24E+12	1.24E+12	2.78E+02	1.48E+02	0	0	0
0.75	0	0	6.41E+03	5.13E+04	5.13E+04	1.10E+12	1.10E+12	2.47E+02	1.42E+02	0	0	0
0.8		0	0	6.15E+03	4.53E+04	4.53E+04	9.73E+11	9.73E+11	2.18E+02	1.36E+02	0	0	0
0.85	0	0	5.89E+03	3.98E+04	3.98E+04	8.54E+11	8.54E+11	1.91E+02	1.30E+02	0	0	0
0.9		0	0	5.63E+03	3.47E+04	3.47E+04	7.46E+11	7.46E+11	1.67E+02	1.25E+02	0	0	0
0.95	0	0	5.37E+03	3.01E+04	3.01E+04	6.47E+11	6.47E+11	1.45E+02	1.19E+02	0	0	0
1		0	0	5.11E+03	2.59E+04	2.59E+04	5.58E+11	5.58E+11	1.25E+02	1.13E+02	0	0	0



**Note: If this file is for a TOWER, the following section properties are read but overwritten as follows:
  str_tw is set to zero
  tw_iner is set to zero
  cg_offst is set to zero
  sc_offst is set to zero
  tc_offst is set to zero
  edge_iner is set equal to flp_iner
  edge_stff is set equal to flp_stff
`

**Reply by jjonkman** [reply]
Dear @EWEMkite,

Are you trying to calculate the tower mode shapes for a specific reference model provided by NREL, such as the WindPACT 1.5-MW baseline, the NREL 5-MW baseline, or the IEA Wind 15-MW RWT?  I see few references in your input files, but not sure which one is correct.

Best regards,

**Reply by EWEMkite** [reply]
Dear @jjonkman,

The reference model is the IEA 15MW RWT.

**Reply by jjonkman** [reply]
Dear @EWEMkite,

I would guess the reason for the very low natural frequencies you are getting from BModes is because your `tor_stff` and `axial_stff` values are much lower than I would expect them to be.

FYI: While these are not mine, BModes input files for the IEA Wind 15-MW RWT atop the monopile are available on our forum here: https://forums.nrel.gov/t/input-parameters-of-tower-for-bmodes-and-elastodyn-iea-15-mw/4151.  I haven't cross checked all of your numbers, but I certainly noticed differences in multiple values.

Best regards,

**Reply by EWEMkite** [reply]
Dear @jjonkman,

Thanks for the help, the frequency now looks more correct. But when computing the output of BModes to get the mode shape x^2~x^6 coefficients for ElastoDyn_Tower, I think the values are suspicious, could you please help me have a check?

My model is a floating wind turbine system, the floater is a self-designed spar buoy, the turbine is IEA-15MW turbine and the mooring system is copied from WindCrete (which is also a spar buoy that supports the IEA-15MW turbine) input file. I read through the posts about BModes setting on the forum, and set some key values in the input file by:

1. Platform mass inertia 3X3 matrix (i_matrix_pform):
 I found that this the same as the ElastoDyn setting, when comparing the input files of the OC3 Hywind model. So I take my values from ElastoDyn as input here, and they are contributed by the platform (the floater and the ballast) which doesn't include tower, RNA, etc as discussed on other posts.
2. Blade-tip or tower-top mass properties:
I refer to this post[https://forums.nrel.gov/t/input-parameters-of-tower-for-bmodes-and-elastodyn-iea-15-mw/4151](url)
3. tow_support:
I tried with 2, but the version of BModes_jj only allows 0 or 1, I checked the OC3Hywind.bmi and found that it's also set as 1, so I set this number as 1, too.
4. Platform-reference-point-referred hydrodynamic 6X6 matrix (hydro_M):
Scaled from WAMIT .1 output.
5. Platform-reference-point-referred hydrodynamic 6X6 stiffness matrix (hydro_K):
Scaled from WAMIT .hst output, and (6,6) is augmented with the additional yaw stiffness which is found on the description of WindCrete mooring line:
![image](https://github.com/user-attachments/assets/d2e58abe-90a9-4ca7-a3db-970a579a998e)
6. Mooring-system 6X6 stiffness matrix (mooring_K):
Generated by MAP++, (4,4) and (5,5) are augmented by -mgz, m and z refer to the whole system

I have attached my inputs and outputs, I took the mode shapes of 7th-10th for calculations as I saw on a post that the 1st~6th mode refer to the platform 6dofs. From my understanding, the output file of BModes has:
The 7th mode refers to the tower s-s first mode;
The 8th mode refers to the tower f-a first mode;
The 9th mode refers to the tower s-s first mode;
The 10th mode refers to the tower f-a first mode;

But I am not sure why in my output file the 10th mode has 0 f-a disp values. Additionally, I noticed that these disp values are modal displacement values. What is the relation between these modal displacement values and the displacement with unit of [m]?  

Best Regards.

[IEA15_CSpar_Bmodes.zip](https://github.com/user-attachments/files/19044487/IEA15_CSpar_Bmodes.zip)

**Reply by EWEMkite** [reply]
Sorry, just to correct a bit:

The 7th mode refers to the tower s-s first mode;
The 8th mode refers to the tower f-a first mode;
The 9th mode refers to the tower s-s second mode;
The 10th mode refers to the tower f-a second mode;

**Reply by jjonkman** [reply]
Dear @EWEMkite,

Just a few comments:

- While I haven't checked all of your BModes inputs, overall your approach sounds OK, although I'm not sure I understand what the additional yaw stiffness is in your step 5.
- Do the natural frequencies reported for modes 1-6 match what you expect the rigid-body floater modes (surge, sway, heave, roll, pitch, yaw) to have?
- I agree that modes 7 and 8 correspond to the 1st tower side-side and fore-aft bending modes.
- Mode 9 is not the 2nd tower side-side bending mode; rather, it appears to be the 1st tower torsion mode.
- Mode 10 looks more like the 2nd tower side-side bending mode, which probably means Mode 11 (which you haven't output) corresponds to the 2nd tower fore-aft bending mode.

Best regards,

**Reply by EWEMkite** [reply]
Dear @jjonkman,

Thanks for clarifing, I haven't checked the natural frequencies, will it be suitable to use the natural frequencies found from WAMIT outputs (the RAOs-frequency peak points) to compare? 

As for the additional yaw stiffness, this refers to this post: https://forums.nrel.gov/t/bmodes-input-parameters-about-tower-support-subsystem/604, where you stated that the (6, 6) of hydro_K is augmented with the additional yaw stiffess.
![image](https://github.com/user-attachments/assets/2e618d76-963f-4aa8-b4e5-aeadaccd4668)

Best Regards

**Reply by jjonkman** [reply]
Dear @EWEMkite,

I would not expect tower flexibility to have much effect on the rigid-body floater modes, so, "yes," you should be able to compare the frequencies of these modes to what you see from WAMIT, which assumes a rigid FOWT.

So, the yaw spring you've added is used to capture the effect of the delta connection between the spar and mooring lines?  FYI: We previously added a yaw spring because old versions of FAST could not model the delta connection directly.  In MAP++ within OpenFAST, the delta connection can be modeled directly, so, there should be no need to add a yaw spring if you are modeling the delta connection within MAP++.

Best regards,

**Reply by EWEMkite** [reply]
Dear @jjonkman,

Thanks! These info are really helpful. Here is my MAP file that I used to generate the mooring stiffness matrix:

--------------- LINE DICTIONARY ----------------------------------------------
LineType  Diam      MassDenInAir   EA        CB   
(-)       (m)       (kg/m)        (N)        (-)  
CHAIN     0.160    561.2520     2.3040e+09      1.0  
--------------- NODE PROPERTIES ----------------------------------------------
Node Type       X       Y       Z         M     B     FX    FY    FZ
(-)  (-)       (m)     (m)     (m)       (kg)  (mˆ3)  (N)   (N)   (N)
1    fix   -600.0      0.0      -200      0     0     #     #     #     
2  connect #-47.665   #0.0   #-114.23     0     0     0     0     0   
3    fix    300    -519.6152   -200       0     0     #     #     #    
4  connect #23.8325  #-41.2791  #-114.23    0     0     0     0     0
5   fix     300   519.6152   -200       0     0     #     #     #   
6   connect #23.8325  #41.2791  #-114.23    0     0     0     0     0
7   vessel    -4.65    8.054   -90.00      0     0     #     #     #
8   vessel    -4.65   -8.054   -90.00      0     0     #     #     #
9   vessel     9.3     0.0    -90.00      0     0     #     #     #   
10   vessel    -4.65    8.054   -90.00      0     0     #     #     #
11   vessel    -4.65   -8.054   -90.00      0     0     #     #     #
12   vessel     9.3     0.0    -90.00      0     0     #     #     #   
--------------- LINE PROPERTIES ----------------------------------------------
Line    LineType  UnstrLen  NodeAnch  NodeFair  Flags
(-)      (-)       (m)       (-)       (-)       (-)
1        CHAIN      565      1         2         TENSION_ANCH ALTITUDE_ANCH 
2        CHAIN      50.0      2         7         TENSION_FAIR 
3        CHAIN      50.0      2         11         TENSION_FAIR 
4        CHAIN      565      3         4         TENSION_ANCH ALTITUDE_ANCH
5        CHAIN      50.0      4         8         TENSION_FAIR
6        CHAIN      50.0      4         12         TENSION_FAIR
7        CHAIN      565      5         6         TENSION_ANCH ALTITUDE_ANCH
8        CHAIN      50.0      6         10         TENSION_FAIR
9        CHAIN      50.0      6         9         TENSION_FAIR
--------------- SOLVER OPTIONS------------------------------------------------
Option
(-)
outer_tol 1e-2

As there are 'connect' node types, I suppose the effect of the delta connection is already taken into account?

Additionally, except for WAMIT, I am curious that is there any other suggested way to calculate and check the natural frequency with BModes? In my mind I only know decay tests but they require these mode shape coefficients as inputs.. 

Best Regards,
Thanks for the fast response.

**Reply by jjonkman** [reply]
Dear @EWEMkite,

I agree that your MAP++ is already modeling the delta connection, so, there is no need to add a yaw spring.

You identify the rigid-body mode natural frequencies from the OpenFAST model without the tower degrees of freedom enabled, in which case the tower mode shapes will not affect the solution.

Best regards,

**Reply by EWEMkite** [reply]
Dear @jjonkman,

I noted the natural frequencies down from the results of BModes and decay tests, except for heave and pitch frequency values don't match, other dofs look fine:
![image](https://github.com/user-attachments/assets/e2747f6a-67d6-4d11-af75-ab6c6775d697)

The values of heave and pitch dofs seem 'swapped', I assume that in BMode outputs dof 3 is heave and dof 5 is pitch, the BMode output file is attached. I am not sure why the values are swapped in these two approaches. 

An additional question is, as I found that the tower frequencies are very close to my reference turbine IEA 15MW's 3P lower bound (0.25Hz), I want to know if the hull modelled in BModes_jj is flexible or rigid, as I read that the frequencis outputs could be actually lower if the hull is modeled as rigid. 

Best Regards
[IEA15_CSpar_Bmodes_bmi.txt](https://github.com/user-attachments/files/19554437/IEA15_CSpar_Bmodes_bmi.txt)
[IEA15_CSpar_Bmodes_out.txt](https://github.com/user-attachments/files/19554448/IEA15_CSpar_Bmodes_out.txt)

**Reply by jjonkman** [reply]
Dear @EWEMkite,

I haven't reviewed your BModes input file, but I would say you swapped the order of modes in your intepretation of the BModes output file.  Mode 3 is roll (with roughly constant side-side slope), Mode 4 is pitch (with roughly constant fore-aft slope), and Mode 5 is heave (that is not really possible to "see" in the BModes outputs, that only report side-side, fore-aft, and twist deflections).  With these interpretations, BModes matches your decay tests well.

Best regards,

**Reply by EWEMkite** [reply]
Dear @jjonkman,

my decay tests matched well with results of BModes as shown in the previous posts, but recently I add SubDyn into the simulation and redo the decay tests, I found the results are now very different but not sure why.

As suggested from other posts I found, I have the following settings (take heave as example):
![image](https://github.com/user-attachments/assets/d914e1e7-081c-4264-819a-f06a063cd81f)
![image](https://github.com/user-attachments/assets/9a47342e-e236-4ca6-9cab-e879bc341086)

While BModes_jj gives me heave frequency of 0.040747:
![image](https://github.com/user-attachments/assets/9b0256c5-8723-4707-a749-6030750a08b3)
The decay test presents a much smaller value, around 1/120=0.008Hz:
![image](https://github.com/user-attachments/assets/9b76756f-8859-4c80-8dc1-807361614c8f)

As BModes only has tower and platform part, I set all the platform properties in BModes to include Floater Structure and Ballast. In the decay tests I run with openFAST however, I set the platform properties with only the ballast mass, while the floater is modeled with SubDyn. My consideration here is according to this post (I think this post refers to openFAST model with only ElastoDyn and without SubDyn):
![image](https://github.com/user-attachments/assets/65987aff-89fa-4705-b8cc-a66647d168a7)

I am not sure if I set anything wrong with SubDyn or BModes that leads to the difference. Could you please help me with this?

Best Regards,

**Reply by jjonkman** [reply]
Dear @EWEMkite,

Well, I would not expect switching from modeling a rigid floater in ElastoDyn to a flexible floater in SubDyn to greatly impact the rigid-body natural frequencies.  So, I would question whether your model with SubDyn is set up properly.

When you toggle SubDyn on / off, are you removing the same mass from ElastoDyn that you are adding in SubDyn? (The integrated mass implemented in the SubDyn model is reported in the SubDyn summary file.)   You could also try keeping SubDyn enabled, but making it rigid (keeping only the rigid-body Guyan modes), by setting `Nmodes` = 0 in SubDyn.

Best regards,

**Reply by EWEMkite** [reply]
Dear @jjonkman ,

I removed the same mass from ElastoDyn to SubDyn yes, and the integrated mass also matched my static calculation (Buoyancy = Gravity + Mooring Pre-tension): 
4.094086E+08 = (mass_ED + mass_SD)g + 7.37E6

![image](https://github.com/user-attachments/assets/f537b742-bdcc-4fa5-9077-0d8c34618ba3)
![image](https://github.com/user-attachments/assets/d8a389a0-89ef-47fe-aff7-6d3a0f7cba50)
![image](https://github.com/user-attachments/assets/95442d03-90e6-4a93-ba0d-24ebee1b6c47)

the model is also not sinking or lifting according to the motions, but the natural periods are just much longer than BModes predicts. 

I tried with Nmodes=0 in Subdyn but the periods are still much longer. My SubDyn settings are as follows:
![image](https://github.com/user-attachments/assets/cd56ed32-8448-4976-bc38-82170b21e9df)
![image](https://github.com/user-attachments/assets/66b0a232-ab50-4df5-a59c-ad0dae7cfc6e)

**Reply by jjonkman** [reply]
Dear @EWEMkite,

Can you clarify what issue you are seeing?  When you toggle SubDyn on / off in OpenFAST, and toggle `Nmodes` = 0 or higher in SubDyn, how do the three OpenFAST solutions compare?

Best regards,

**Reply by EWEMkite** [reply]
Dear @jjonkman,

Case 1: Subdyn on with Nmodes=10,  very small natural frequency compared to Bmodes outputs. You can see that from https://github.com/OpenFAST/openfast/discussions/2551#discussioncomment-13094892 the heave decay figure has natural period of around 120s, while the Bmodes outputs about 25s (0.04Hz) for heave (mode 5)

Case 2: Subdyn on with Nmodes=0, still very small natural frequency same as case 1.

Case 3: No Subdyn, the substructure (floater) is modeled in ElastoDyn. The decay tests match Bmodes outputs. These refer to the previous post https://github.com/OpenFAST/openfast/discussions/2551#discussioncomment-12690832

I suppose there’s either stiffness missing at the substructure part (maybe hydro stiffness?) or mass missing, but according to the mass balance I think it’s more like missing stiffness. As the hydrodynamic properties are defined in HydroDyn I am not sure if SubDyn take those hydrodynamics into account.

Best regards

**Reply by jjonkman** [reply]
Dear @EWEMkite,

OK, so OpenFAST is not agreeing when you toggle SubDyn on/off, regardless if the substructure is modeled rigidly (`Nmodes` = 0) or not (`Nmodes` > 0).  For Case 2 with `Nmodes` = 0 in SubDyn, can you also try disabling the Static Improvement Method by setting `SttcSolve` = FALSE; this should make SubDyn completely rigidly, without any internal stiffness effect and I would think the solution should match Case 3 without SubDyn.  Can you confirm?

Have you tried to visualize the Guyan and Craig-Bampton modes of the SubDyn model through the following website created by Emmanuel Branlard: [https://ebranlard.github.io/viz3Danim/](https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Febranlard.github.io%2Fviz3Danim%2F&data=05%7C02%7Cjason.jonkman%40nrel.gov%7C10370edaea694c31a27308dd162f7f07%7Ca0f29d7e28cd4f5484427885aee7c080%7C0%7C0%7C638691115434866939%7CUnknown%7CTWFpbGZsb3d8eyJFbXB0eU1hcGkiOnRydWUsIlYiOiIwLjAuMDAwMCIsIlAiOiJXaW4zMiIsIkFOIjoiTWFpbCIsIldUIjoyfQ%3D%3D%7C0%7C%7C%7C&sdata=v37%2B7lKwnyJlc3Xx8ljFiViCtuQwFeGKHo90XnDY56w%3D&reserved=0); to use this feature, simply load the _.json_ summary file generated by SubDyn.  Do the SubDyn Guyan and Craig-Bampton mode shapes make physical sense, e.g., are the Guyan modes rigid body modes in your SubDyn model?

Best regards,

**Reply by EWEMkite** [reply]
Dear @jjonkman,

Thanks for the response, these two days I checked the .hst file and I found somehow it's set all 0 values, which means there was no hydrostatic restorings. After I corrected it the responses now match the ones I got before and also BModes outputs.

I will try your advice and write different settings in SubDyn in the next step to see how the mode shapes look like.

Best Regards
